src/inference/gibbs.jl

```diff
@@ -62,11 +78,117 @@ function Sampler(alg::Gibbs, model::Model, s::Selector)
         @warn("[Gibbs] extra parameters specified by samplers don't exist in model: $(setdiff(space, get_pvars(model)))")
     end
 
-    info[:samplers] = samplers
+    # Create a state variable.
+    state = GibbsState(model, samplers)
+
+    # Create the sampler.
+    spl = Sampler(alg, info, s, state)
+
+    # Add Gibbs to gids for all variables.
+    for sym in keys(spl.state.vi.metadata)
+        vns = getfield(spl.state.vi.metadata, sym).vns
+        for vn in vns
+            # Update the gid for the Gibbs sampler.
+            Turing.RandomVariables.updategid_forced!(spl.state.vi, vn, spl)
+            
+            # Try to store each subsampler's gid in the VarInfo.
+            for local_spl in spl.state.samplers
+                Turing.RandomVariables.updategid!(spl.state.vi, vn, local_spl)
+            end
+        end
+    end
 
     return spl
 end
 
+# First step.
```

Comment:
Why do we need 2 steps here?

Reply:
I answer this at the bottom, but I wanted to provide some background first. There are two general signatures for `step!`, one of which is called for the very first sample and one of which is called for the second. This is because each `step!` function has the ability to depend on a past sample draw, but a past sample draw does not exist in the first step.

From the `sample` function:

```julia
    # Step through the sampler.
    for i=1:N
        if i == 1
            ts[i] = step!(rng, ℓ, s, N; iteration=i, kwargs...)
        else
            ts[i] = step!(rng, ℓ, s, N, ts[i-1]; iteration=i, kwargs...)
        end

        # Run a callback function.
        callback(rng, ℓ, s, N, i, cb; kwargs...)
    end
```

This `step!` function accepts a `Nothing` in the last argument, and it is only called the first time. For some samplers this is important -- the MH sampler for example always accepts whatever the first draw is. For most samplers it doesn't matter, and in those cases the sampler just defaults to a single `step!` function that doesn't care about receiving a `Nothing`.

```julia
function step!(
    rng::AbstractRNG,
    ℓ::ModelType,
    s::SamplerType,
    N::Integer,
    t::Nothing;
    kwargs...
)
```

This next `step!` function is called the most, and it includes some past observation of a `Transition` object.

```julia
function step!(
    rng::AbstractRNG,
    ℓ::ModelType,
    s::SamplerType,
    N::Integer,
    t::TransitionType;
    kwargs...
) 
```

To answer your question, we need two steps so that each `Gibbs` step can pass on the signature to the component samplers where it matters. For HMC, it doesn't matter -- it'll just use the same step function regardless. But if you are using HMC and MH, it really matters that the first step has a different functionality for MH. This setup captures that and allows each sampler to dispatch accordingly.

Reply:
Can you remind me why the first draw of MH is always accept?